text/0000-async-stream.md

```diff
@@ -0,0 +1,883 @@
+
+In addition to adding the `Stream` trait to the standard library, we also want to provide basic 
+ergonomic methods required to use streams effectively. These include the
+`next` and `poll_next` methods. Without these methods, `Streams` would feel much more 
+difficult to use. Were we to not include them, users would likely immediately reach out 
+for them either through writing their own version or using an external crate. 
+
+Including `Stream` in the standard library would also clarify the stability guarantees of the trait. For example, if [Tokio](https://tokio.rs/) 
+wishes to declare a [5 year stability period](http://smallcultfollowing.com/babysteps/blog/2020/02/11/async-interview-6-eliza-weisman/#communicating-stability), 
+having the `Stream` trait in the standard library means there are no concerns 
+about the trait changing during that time ([citation](http://smallcultfollowing.com/babysteps/blog/2019/12/23/async-interview-3-carl-lerche/#what-should-we-do-next-stabilize-stream)).
+
+## Examples of current crates that are consuming streams
+
+### async-h1
+
+* [async-h1](https://docs.rs/async-h1)'s server implementation takes `TcpStream` instances produced by a `TcpListener` in a loop.
+
+### async-sse
+
+* [async-sse](https://docs.rs/async-sse/) parses incoming buffers into a stream of messages.
+
+## Why a shared trait?
+
+We eventually want dedicated syntax for working with streams, which will require a shared trait. 
+This includes a trait for producing streams and a trait for consuming streams.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A "stream" is the async version of an [iterator].
+
+The `Iterator` trait includes a `next` method, which computes and returns the next item in the sequence. The `Stream` trait includes two core methods for defining and interacting with streams - `poll_next` and 'next`.
+
+## poll_next method
+
+
+When implementing a `Stream`, users will define a `poll_next` method. 
+The `poll_next` method asks if the next item is ready. If so, it returns
+the item. Otherwise, `poll_next` will return [`Poll::Pending`]. 
+
+Just as with a [`Future`], returning [`Poll::Pending`] 
+implies that the stream has arranged for the current task to be re-awoken when the data is ready.
+
+[iterator]: https://doc.rust-lang.org/std/iter/trait.Iterator.html
+[`Future`]: https://doc.rust-lang.org/std/future/trait.Future.html
+[`Poll::Pending`]: https://doc.rust-lang.org/std/task/enum.Poll.html#variant.Pending
+
+```rust
+// Defined in std::stream module
+pub trait Stream {
+    // Core items:
+    type Item;
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>>;
+    
+    // Optional optimization hint, just like with iterators:
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (0, None)
+    }
+
+    // Convenience method (covered later on in the RFC):
+    fn next(&mut self) -> Next<'_, Self>
```

Comment:
I wonder, could this be `-> impl Future<Output = Option<Self::Item>>` instead? That would allow us to not introduce `Next` as a public type.